### PR TITLE
[core][state][log] State API should not allow reading files outside of the ray log directory on all ray nodes.

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
State API log retrieval has a bug where one could pass:
1. relative paths like "../../../xxx" to get file outside of ray's log dir
2. absolute path that's refers to other files to get file somewhere else. 

This PR fixes both issues such that one could only read logs under the ray logs directory.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
add a unit test with non-log folder directory access?

File: dashboard/modules/log/log_agent.py [hunk 1]
```diff
@@ -370,6 +374,35 @@ async def _find_task_log_offsets(
...
+        else:
+            filepath = Path(request.log_file_name)
+
+        filepath = filepath.resolve()
+
+        if not filepath.is_file():
+            raise FileNotFoundError(f"A file is not found at: {filepath}")
+
+        try:
+            filepath.relative_to(self._dashboard_agent.log_dir)
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply dashboard/modules/log/log_agent.py:398]:
```
>>> a = Path("/tmp")
>>> log_root = Path("/tmp/ray/session_latest")
>>> a.relative_to(log_root)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/home/rickyx/anaconda3/envs/dev/lib/python3.8/pathlib.py", line 908, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: '/tmp' does not start with '/tmp/ray/session_latest'
```

**Review by rkooo567 — APPROVED** [review]
cc @thomasdesr for the final sign off

**Review by thomasdesr — APPROVED** [review]
Looks great, thanks :D

File: dashboard/modules/log/log_agent.py [hunk 2]
```diff
@@ -370,6 +374,38 @@ async def _find_task_log_offsets(
...
+
+        filepath = filepath.resolve()
+
+        if not filepath.is_file():
+            raise FileNotFoundError(f"A file is not found at: {filepath}")
+
+        try:
+            filepath.relative_to(root_log_dir)
+        except ValueError as e:
+            raise FileNotFoundError(f"{filepath} not in {root_log_dir}: {e}")
```
Comment (thomasdesr) [inline dashboard/modules/log/log_agent.py:405]: Out of curiosity, any reason we're going the exception throwing path vs `is_relative_to`?

Comment (pcmoritz) [reply dashboard/modules/log/log_agent.py:405]: I had the same comment (@rickyyx feel free to make a follow up PR to change this, but it is cosmetic so I'm merging the PR which is an important fix)

Comment (rickyyx) [reply dashboard/modules/log/log_agent.py:405]: oh - I think that's only available in py3.9?
https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to

Comment (thomasdesr) [reply dashboard/modules/log/log_agent.py:405]: Oooo, very good catch. Thanks!

**Pull request merged**